### ISSUE TITLE
Avoid transitive dependencies in module-info

### ISSUE THREAD

**Issue by Sanne** [eclipse-ee4j/jpa-api#264]
Hello,
many thanks for introducing a module-info, that's great progress.

I wonder if it would be possible to apply some minor tuning; in particular I would prefer, if possible, if you could remove the `transitive` qualifier from all modules.

For context, I work on the Hibernate project, and while I understand this might be surprising, we actually do have several projects which don't necessarily make use of neither `java.instrument` nor `java.sql`, while still using the JPA API.

I would suggest that importing of such modules should be left to the implementors?

Tangentially, we don't use `java.instrument` and I noticed the JPA API only requires it for the exception on a signature from `jakarta.persistence.spi.ClassTransformer` - which in several runtimes we don't use. So ideally it would be great to be able to use an optimised JDK which doesn't load the `java.instrument` module at all, but I realize this is an existing API and we can't do much about it yet.

Happy to send a PR to remove the `transitive` qualifier, if you agree?

Thanks!

**Comment by scottmarlow** [comment]
Thanks for raising this @Sanne!  I was just looking at [module-info](https://github.com/eclipse-ee4j/jpa-api/blob/master/api/src/main/java/module-info.java).  

How about java.logging?

I wonder if the module-info.java source document via comments, the dependencies instead:
`
/* requires java.logging;
    requires transitive java.instrument;
    requires transitive java.sql; */
`

**Comment by Sanne** [comment]
> How about java.logging?

Hibernate ORM doesn't that either, but this one isn't marked as `transitive` it's not a problem: I'd say that makes it an internal implementation detail.

Clearly I'd be even happier if it wasn't depending on it but since we're not here to debate internal implementation details I won't complain :)

**Comment by anthonyvdotbe** [comment]
Both modules appear in the public API (e.g. `java.instrument` in javax.persistence.spi.ClassTransformer and `java.sql` in javax.persistence.spi.PersistenceUnitInfo), so I believe they should be marked as transitive (if I'm not mistaken, javac will also give an Xlint warning if they're not).

The modular solution would be to define a "JPA lite" which doesn't depend on those modules, and redefine the API module as an aggregate module:
````java
module jakarta.persistence {
    requires transitive jakarta.persistence.core;
    requires transitive jakarta.persistence.spi;
}
````

**Comment by Sanne** [comment]
yes, I relize they are transitive so to make it easier to consumers - but I'm one of those consumers and I'm not going to need them, so I'd rather see it as an explicit opt-in we should do.

After all it's extremely simple to add it, and the compiler will tell you right away that it's missing (if you forgot) - but there is no way to remove them.

**Comment by Sanne** [comment]
> I believe they should be marked as transitive

I would generally agree with you as a general guideline, but technically most user (the end user application) will really not need these at all. I'm actually pretty sure that in some places it wouldn't be allowed to let the business logic layer access the `java.instrument` module.

> (if I'm not mistaken, javac will also give an Xlint warning if they're not).

I might be looking in the wrong place, but running a standard maven build I don't see any such warnings.

**Comment by anthonyvdotbe** [comment]
> I might be looking in the wrong place, but running a standard maven build I don't see any such warnings.

This is because the maven-compiler-plugin isn't configured to show warnings (with `<showWarnings>true</showWarnings>`). If it is, and the `transitive` qualifiers are removed, it'll print:

````
[WARNING] .../jpa-api/api/src/main/java/jakarta/persistence/criteria/CriteriaBuilder.java:[1224,24] class java.sql.Date in module java.sql is not indirectly exported using requires transitive
[WARNING] .../jpa-api/api/src/main/java/jakarta/persistence/criteria/CriteriaBuilder.java:[1230,24] class java.sql.Timestamp in module java.sql is not indirectly exported using requires transitive
[WARNING] .../jpa-api/api/src/main/java/jakarta/persistence/criteria/CriteriaBuilder.java:[1236,24] class java.sql.Time in module java.sql is not indirectly exported using requires transitive
[WARNING] .../jpa-api/api/src/main/java/jakarta/persistence/spi/PersistenceUnitInfo.java:[68,12] interface javax.sql.DataSource in module java.sql is not indirectly exported using requires transitive
[WARNING] .../jpa-api/api/src/main/java/jakarta/persistence/spi/PersistenceUnitInfo.java:[80,12] interface javax.sql.DataSource in module java.sql is not indirectly exported using requires transitive
[WARNING] .../jpa-api/api/src/main/java/jakarta/persistence/spi/ClassTransformer.java:[61,16] class java.lang.instrument.IllegalClassFormatException in module java.instrument is not indirectly exported using requires transitive
````

**Comment by Sanne** [comment]
thanks for checking that @anthonyvdotbe , appreciate it!

Still, let's look at the experience that the JPA _user_ will have (not the implementor): why should they have access to the `java.instrument` and `java.sql` module just to annotate their domain module - and possibly use the `EntityManager` ?

One could theoretically argue that `java.sql` could be useful for a JPA user, but giving access to instrumentation seems to be a violation of the encapsulation goal of the modules system.

**Comment by anthonyvdotbe** [comment]
Yes, I agree that it's very unlikely that a JPA user would actually notice the difference. And I believe that removing the `transitive` qualifiers is an acceptable short-term solution, but that the module should be properly split into submodules in the long run.

W.r.t. `why should they have access` and `giving access to instrumentation seems to be a violation of the encapsulation goal of the modules system`, I'm not sure I understand your concern: users could just add `requires java.instrument` in their own module descriptors, no? Encapsulation is primarily about preventing access to unexported packages.

Just for the record: I'm not a JPA contributor or anything, so these are just my 2 cents.

**Comment by Sanne** [comment]
right, splitting the module might make sense - however keep in mind that at this stage we're not allowed to make any other API change other than the package name.

So while those are good suggestions for the future, for the time being I'd be happy enough to just remove the `transitive`.

Alternatively perhaps the whole module-info should be removed; I didn't expect that to be introduced as it feels like a change which goes quite a bit beyond the package rename.

**Comment by lukasj** [comment]
I do not agree with entirely removing `transitive` from both places for following reasons:

- as @anthonyvdotbe already said - both modules appear in the public API
- `spi` is integral part of existing spec and API implementation jar should follow that

wrt `java.instrument`:
- dependency on what has become `java.instrument` module has been added in 2008 (some 11,5 years ago). I doubt that at that time anyone knew how java module system will look like.
- be it just on me, I would prefer to use either some JPA API specific exception or generic `PersistenceException` instead even at that time just to avoid this direct cross-package/technology dependency and to keep doors open wider for the future. I'm afraid that it is too late to ask for this kind of breaking API change now
- I can agree changing `requires transitive java.instrument` to `requires static java.instrument` as that is what better reflects the optionality of implementation of `ClassTransformer` interface by the provider[1] and partially solves your problem.

I do not agree with complete removal of `module-info` as that would not allow one to use an optimized JDK (`jlink` does not support automatic modules) and it seems to me that one wants _to be able to use an optimised JDK which doesn't load the ~~java.instrument~~_ unnecessary _modules at all_. Existing specification as such does not explicitly forbid nor mandates its existence and its presence as such can be seen as an implementation detail.

[1]: section 9.6.1 of the spec:

> The javax.persistence.spi.ClassTransformer interface is implemented by a persistence
provider that wants to transform entities and managed classes at class load time or at class redefinition
time.

**Comment by bshannon** [comment]
I agree with @anthonyvdotbe.  These types are part of the JPA API so they
need to be declared in the module-info.  And yes, if you don't like that, carve
up JPA into spec-defined subsets that don't include these dependencies.

**Comment by keilw** [comment]
The idea for a `jakarta.persistence.core` is really great, even if it may not happen in Jakarta EE 9 yet, because that way we could explore possible synergies with Jakarta NoSQL and try to reuse some parts across both.

**Comment by lukasj** [comment]
I've tried to make dependency on `java.instrument` explicitly optional through #264. Opinions welcomed.

**Comment by lukasj** [comment]
took silence as an agreement

**Comment by jgrassel** [comment]
I'm a bit concerned about the "silence as an agreement" approach, as the last bit of dialog about this issue was 8 months ago and likely was forgotten about everyone.  I think that conversation should have been renewed here or posted on the `jpa-dev@eclipse.org` mailing list to determine that there was sufficient support for this change.  I was a bit surprised to see any feature additions added to the JPA 3.1 spec since the `3.1-3.1.0-RC1-RELEASE` tag was cut, and was expecting any new commits to be only bugfixes to previous commits, and not new issues.  Just my 2 cents.

**Comment by lukasj** [comment]
`3.1-3.1.0-RC1-RELEASE` means *almost* nothing as every build from the spec project repository is marked as RC _by default_. The real purpose of these builds is to allow integrations and testing in implementations.

**Comment by jgrassel** [comment]
Gotcha, if we have the means to change it, we should.  Something like "MS" (for Milestone) if these are tags produced by the build system.  RC sounds like "Release Candidate", which is definitely not the same thing.

**Comment by dazey3** [comment]
I find it kind of odd that this API breaking change is targeted for a minor version update of the JPA specification. I did not expect `jakarta.persistence.spi.PersistenceUnitInfo.addTransformer(ClassTransformer transformer)` to be completely removed until the next major version update and instead expected a simple deprecation. 

The `jakarta.persistence.spi.ClassTransformer` API was deprecated though, which makes the decision to completely remove the `jakarta.persistence.spi.PersistenceUnitInfo.addTransformer(ClassTransformer transformer)` API method even more confusing.

**Comment by lukasj** [comment]
@dazey3 

> this API breaking change

where exactly do you see the incompatibility?

was:
```
addTransformer(ClassTransformer transformer)
```

is:
```
ClassTransformer extends Transformer
addTransformer(Transformer transformer)
```

so the new `addTransformer(Transformer transformer)` works perfectly fine for old `ClassTransformer`s.

**Comment by jgrassel** [comment]
It's a method signature change to `PersistenceUnitInfo.addTransformer()`, where its single argument type was `ClassTransformer` but is now `Transformer`.  This is breaking backwards compatibility with previous versions of the SPI -- point releases are typically additive, with subtractive changes being major version (4.0) boundary changes.  It's a lot more significant than adding a UUID enum value IMHO.

**Comment by dazey3** [comment]
@lukasj 

> > this API breaking change

> where exactly do you see the incompatibility?

Correction, I should have said SPI breaking change, not API breaking.

**Comment by lukasj** [comment]
right, method signature is changed. Do you have an example of an existing code which does not compile and/or run without modifications with updated SPI? ClassTransformer extends Transformer, therefore it is safe to pass existing instance of the ClassTransformer to the new addTransformer(Transformer transformer) method and the transform method is still there and unchanged in the ClassTransformer => compiler won't complain; the new Transformer contains new transform method (there is a difference in the order of first two arguments there)

**Comment by lukasj** [comment]
No, we are not broken.

**Comment by dazey3** [comment]
JPA 3.0.0

```java
public class MyPersistenceUnitInfo implements jakarta.persistence.spi.PersistenceUnitInfo {

	@Override
	public void addTransformer(ClassTransformer arg0) {
		// implementation
	}
}
```

Will this compile with the new SPI?

**Comment by jgrassel** [comment]
>  ClassTransformer extends Transformer, therefore it is safe to pass existing instance of the ClassTransformer to the new addTransformer(Transformer transformer) 

Except now JPA provider implementors can now pass objects which implement Transformer without also implementing ClassTransformer, existing code (such as appservers and libraries like presumably Spring) which utilizing PersistenceUnitInfo can no longer assume that the object will always be a ClassTransformer (hello ClassCastExceptions).  That's a very significant change for a point release.

**Comment by scottmarlow** [comment]
Agreed that JPA containers will need to choose between calling the persistence provider Transformer#transform versus the ClassTransformer#transform method depending on whether PersistenceUnitInfo.addTransformer was passed a Transformer or ClassTransformer. 

This doesn't seem difficult to me but I guess there could be more to it than I understand now.

**Comment by scottmarlow** [comment]
Does changing from throwing an IllegalClassFormatException to TransformerException help in some way?  If no, perhaps the existing transform method signature could be used (e.g. currently classloader + classname parameters are reversed and the throws clause is different).

**Comment by dazey3** [comment]
> JPA containers will need to choose between calling the persistence provider Transformer#transform versus the ClassTransformer#transform method depending on whether PersistenceUnitInfo.addTransformer was passed a Transformer or ClassTransformer.

The issue is not with implementing `javax.persistence.spi.Transformer` vs `javax.persistence.spi.ClassTransformer`. The "Transformer" API was protected for this minor version update by deprecating and extending the interfaces. The PersistenceUnitInfo SPI was broken (meaning that 3.1.0 is no longer backwards compatible with 3.0.0)

**Comment by lukasj** [comment]
We were not using Override in our code for a really long time, that's why I have missed it for the compile time. Let me see if I can find something...

> Does changing from throwing an IllegalClassFormatException to TransformerException help in some way?

yes - it is `java.lang.instrument.IllegalClassFormatException` => java.lang.instrument module while `TransformerException` (or even `PersistenceException` could be used) is in this module (no external dependency)

**Comment by dazey3** [comment]
> We were not using Override in our code for a really long time, that's why I have missed it for the compile time.

I see. So your really just tricking the compiler by saying "well, when the underlying SPI declares the method we override it and when they DONT declare it we just have a public method that no one calls". That certainly might work, but it feels rather hacky.

**Comment by scottmarlow** [comment]
> > Does changing from throwing an IllegalClassFormatException to TransformerException help in some way?

> yes - it is `java.lang.instrument.IllegalClassFormatException` => java.lang.instrument module while `TransformerException` (or even `PersistenceException` could be used) is in this module (no external dependency)

:+1:

**Comment by jgrassel** [comment]
It's been a bit quiet for the last 30 minutes.  Will this PR be rolled back/modified?

**Comment by tjwatson** [comment]
> right, method signature is changed. Do you have an example of an existing code which does not compile and/or run without modifications with updated SPI? ClassTransformer extends Transformer, therefore it is safe to pass existing instance of the ClassTransformer to the new addTransformer(Transformer transformer) method and the transform method is still there and unchanged in the ClassTransformer => compiler won't complain; the new Transformer contains new transform method (there is a difference in the order of first two arguments there)

The problem is with compiled code that is calling the old method taking a ClassTransformer.  The class file will expect to have a `addTransformer(ClassTransformer)` method and you will see a runtime error about the method missing.  Once you recompile the code against the new API it will compile and work fine because the compiler will determine that the addTransformer(Transformer transformer) is valid and compile an invoke to that method signature.  This change means all code must be recompiled that were calling addTransformer.  That is not a minor change.

**Comment by lukasj** [comment]
> > We were not using Override in our code for a really long time, that's why I have missed it for the compile time.
> 
> I see. So your really just tricking the compiler by saying "well, when the underlying SPI declares the method we override it and when they DONT declare it we just have a public method that no one calls". That certainly might work, but it feels rather hacky.

Not really. If Override annotation were present, I'd notice this problem sooner and I'd deal with it differently. The thing is that the existence of the Override annotation is making the change in this area impossible - even if new methods had some default implementations, the possibility of presence of the Override annotation breaks the compilation... 


> It's been a bit quiet for the last 30 minutes. Will this PR be rolled back/modified?

 It's been quiet for months and now it is expected people respond immediately even if they are outside their working hours? Seriously?

(I don't want to be nitpicking but this is even not a PR, this is an issue... ;-) )

**Comment by jgrassel** [comment]
> It's been quiet for months and now it is expected people respond immediately even if they are outside their working hours? Seriously?
> 
> (I don't want to be nitpicking but this is even not a PR, this is an issue... ;-) )

No worries, it's easy to forget about time zones.

Still, this change as it is I think constitutes a major version release, and if we want to keep it as is, we should consider scrapping JPA 3.1, and get the JPA 4.0 project started.

**Comment by lukasj** [comment]
Hm, the simplest solution probably looks like the best one - change the checked exception from  `java.lang.instrument.IllegalClassFormatException` to something else and be done. While that breaks source compatibility, it [does not break binary compatibility](https://docs.oracle.com/javase/specs/jls/se11/html/jls-13.html#jls-13.4.21) and it is likely that it affects only persistence providers

 The project also should be updated to built with source/release 11 [as recommended](https://eclipse-ee4j.github.io/jakartaee-platform/jakartaee10/JakartaEE10ReleasePlan) - need an issue for that...

IMHO it does not make sense to do 4.0 now just because of this as there are more issues breaking compatibility which one wants to see addressed. Let's finish 3.1 first..

**Comment by jgrassel** [comment]
I'm not sure I'm really in favor of breaking even source compatibility.  Is this issue's value, the removal of `java.instrument` as a module dependency, really of such high value that it can't wait until 4.0?

**Comment by lukasj** [comment]
>  Is this issue [..] really of such high value that it can't wait [..]?

For some people - those who filed the issue and/or did some work on it - probably yes, for others probably no. But that is likely true for every issue.

Unless I hear some positive feedback on the change I'll prepare the revert of the change in the near future.

**Comment by Sanne** [comment]
This looks very promising, thank you.

I couldn't test it yet - I hope to be able to do so soon, but since there seems to be urgent need for feedback let me react at higher level:

In principle it's great, as this would allow us to create valid modules for both Hibernate ORM and Hibernate Reactive. Both are able to use instrumentation in certain platforms, but this is not a requirment and in certain contexts people prefer to disable the capability.

In particular it should be allowed to craft a minimal application which has entites instrumented at build time and optimise the distribution via JLink without needing to include the instrumentation code in the runtime; this is also beneficial from a security perspective and in context of compiling to native images with GraalVM and hopefully project Leyden in the future - so I'd say definitely a step in the right direction.

**Comment by dazey3** [comment]
> Hm, the simplest solution probably looks like the best one - change the checked exception from java.lang.instrument.IllegalClassFormatException to something else and be done.

Does this mean a revert of #313 and reimplementation of this new fix strategy?

Is adding the removed declaration of `jakarta.persistence.spi.PersistenceUnitInfo.addTransformer(ClassTransformer transformer)` back a possibility and simply deprecating it until JPA 4.0? Perhaps utilizing default methods as a way to add new functionality to the interface yet ensuring binary compatibility with legacy interface usage?

**Comment by lukasj** [comment]
@Sanne 

I see 3 options:

1. keep it as it is now - this breaks the binary compatibility as @dazey3 and @tjwatson pointed out, therefore I think this is not the best approach to take
2. update the change by removing new interfaces and methods and only change the checked exception in the `ClassTransformer.transform` method from `java.lang.instrument.IllegalClassFormatException` to something else (the new `TransformerException`?) and be done - this breaks the source compatibility but keeps the binary compatibility; with the upgrade to Java SE 11 as the base version for the API, the source compatibility may be, but not necessarily have to be, affected (the Platform will recompile this with release 11 even if nothing is done here anyway)
3. revert the work, do nothing and wait - probably the worst option and those who will want to support minimal possible JPMS runtime will have to find their own way to it, so in the end, we may end up with providers having their own spec-api jars again for the time being

There is still some time - ~ week to resolve this and I personally prefer option 2

**Comment by Sanne** [comment]
I agree option 2 looks reasonable.

Many thanks!

**Comment by dazey3** [comment]
Please, forgive my laymen understanding and questions

@lukasj 
Option 2:
> While that breaks source compatibility, it does not break binary compatibility and it is likely that it affects only those persistence providers, who do implement the ClassTransformer interface.

https://docs.oracle.com/javase/specs/jls/se11/html/jls-13.html#jls-13.4.21

```
13.4.21. Method and Constructor Throws

Changes to the throws clause of methods or constructors do not break compatibility with pre-existing binaries; these clauses are checked only at compile time.
```

This will impact JPA Container implementations that implement `jakarta.persistence.spi.ClassTransformer` as well then, yes? 

From my understanding, this means when we compile the existing JPA Container code: 

- if we compile against the new JPA 3.1.0 binaries, we will fail for incompatibility as the method signatures have changed?
- if we compile against the older JPA 3.0.0 binaries than we will be fine (since they are implemented against 3.0.0), but runtime won't be affected?

If I am correct in my understanding, how is this not a major change as there is no backwards compatibility?

**Comment by dazey3** [comment]
> nothing gets deprecated

Also, I just want to point out that I have no issue with deprecation. That should be the standard behavior when replacing API/SPI methods for minor releases. You deprecate on minor versions and remove on major versions. That's been my whole issue with the original fix: that it completely removed an SPI method and is not a minor change. 

I'll admit that I have not been following along with the specific issue in the OP and how to fix it. But if #313 was a fix to that issue and the only issue is the backwards compatibility, is there not a 4th option to just add back the `jakarta.persistence.spi.PersistenceUnitInfo.addTransformer(ClassTransformer transformer)` and simply deprecate it until JPA 4.0?

**Comment by jgrassel** [comment]
Option 3 might be the "worst option and those who will want to support minimal possible JPMS runtime" but it at least adheres to the spirit of what changes are allowed per major and minor version updates (https://www.eclipse.org/projects/efsp/?version=1.2#efsp-releases-major).  

I wonder if this needs to be brought to next Tuesday's Jakarta EE platform call for resolution.

**Comment by jgrassel** [comment]
I might propose an option 4: keep Transform, deprecate ClassTransformer, mark PersistenceUnitInfo.addTransformer(ClassTransformer) as deprecated (but keep it there, a deprecated item should remain in at least one point release before removal, as it's a warning that it is going to disappear in the future), and add PersistenceUnitInfo.addTransformer(Transformer) with the imperative that the deprecated method is going to go away in 4.0.

This is additive (it doesn't break existing compiles), and gives advanced warning (what deprecation is supposed to do) that it is going away in the next major release, and it doesn't violate the spirit of what is expected of minor version updates.

**Comment by lukasj** [comment]
@dazey3 source compatibility means that compile time is broken but runtime does not care - code compiled with 3.0 will run with 3.1 `as is` but to compile with 3.1 one has to update his code; changing the `IllegalClassFormatException` to its super class (generic `Exception` or even `Throwable`) may make it easier (and possibly preserves source compatibility for implementors you are asking for) - see [Evolving API interfaces](https://wiki.eclipse.org/Evolving_Java-based_APIs_2#Evolving_API_interfaces_-_API_methods) 
 and Weaken/Strengthen Method postconditions in [General Rules for Contract Compatibility](https://wiki.eclipse.org/Evolving_Java-based_APIs#General_Rules_for_Contract_Compatibility). While having generic `Exception` in the API is not nice, it can, IMHO, satisfy your requirement. It can also come with Deprecation, so in the future major release, proper exception can be defined.

Proposed option 4 does not solve the problem since the direct dependency from `PersistenceProvider` through `PersistenceUnitInfo` to `ClassTransformer` and thus also to the  `IllegalClassFormatException` which we want to get rid of/make optional/indirect remains there, so in the end it is the same as current option 3

**Comment by jgrassel** [comment]
Nnngh.  Okay, so I can agree on weakening transform()'s declared exception from `IllegalClassFormatException` to `Exception`, it should be viable.  However, I still have reservations reconciling this against changes expected for major/minor version releases.  Until now, changes have always been additive, I can't say I'm comfortable with this.

**Comment by lukasj** [comment]
I think the main question is: Given that one of the main topics in EE 10 is the Core profile and support for smaller/native runtimes, is there something we can do on our end to make the use of the persistence API easier to use there within current boundaries? I do not think it makes sense to move towards 4.0 just because of this at this stage and 5.0 could show up pretty soon should we want to fix [other incompatibilities](https://github.com/eclipse-ee4j/jpa-api/issues?q=is%3Aopen+is%3Aissue+label%3Aincompatible) having bigger impact. Another item to consider for the next major version could be an update to Generators (Gavin shared some ideas in his PR for UUID support)

**Comment by jgrassel** [comment]
Yeah, I was reading the UUID support issue earlier today, and agreed with the sentiment that it would be better to deliver it all at once instead of piecemeal (so JPA 4.0 for that).  A major release should have substance, and I would agree making 4.0 only about changing this signature would be wasteful (which is why I was advocating for deferring this change to 4.0, and sum it up with other substantive changes).

If one of the goals of EE10 is smaller/native runtimes, then other Jakarta specifications must have run into this same debacle, we can't be the only one.  Can anyone cite how other specs resolved this?

**Comment by lukasj** [comment]
btw: there is also https://eclipse-ee4j.github.io/jakartaee-platform/CompatibilityRequirements - I think this change fits to permitted changes as applications are not supposed to use ClassTransformer directly.

The only similar case I can think of is in Activation and XML Binding which defined dependency on JavaBeans (thus java.desktop module) - both did a major version change to fully get rid of it (JAF 2.0/JAXB 4.0). But in the past there was a "fork" of activation implementation which removed the dependency on java.desktop to allow usage on Android - to answer your question - activation solved it by maintaining an extra fork.

**Comment by jgrassel** [comment]
I wish I could have found that document earlier, it's a lot more helpful than the vague document I found on the eclipse site.  The persistence SPI classes involved in this issue I would say are A) Vendor implementations of the specification, which falls under Permitted Changes #1.  Yeah, I completely agree that the change fits permitted changes as denoted by #1.

**Comment by jgrassel** [comment]
So I am going to change my vote to option 2, as I prefer minimal changes.

**Comment by dazey3** [comment]
> there is also https://eclipse-ee4j.github.io/jakartaee-platform/CompatibilityRequirements - I think this change fits to permitted changes as applications are not supposed to use ClassTransformer directly.

```
These requirements define the kind of changes that are permissible when a Jakarta EE specification project updates a Jakarta EE API specification from one version to the next.
...
1. Impact on vendor implementations: it is expected that a new version of the specification will require existing vendor implementations to be modified and extended to implement the new version.
```

I'm still a bit confused. I don't see anywhere in this document where it is defined what "from one version to the next" means. If this document is trying to say that any time an EE specification releases any sort of new version (like a minor version update in this case) that it is permissible to break the SPI? Can anyone supply a document that describes what "a new version of the specification" is defined as?

According to OSGI versioning semantics: http://docs.osgi.org/whitepaper/semantic-versioning/040-semantic-versions.html

```
major — Packages with versions that have different major parts are not compatible both for providers as well as consumers. For example, 1.2 and 2.3 are completely incompatible.

minor — API consumers are compatible with exporters that have the same major number and an equal or higher minor version. API providers are compatible with exporters that have the same major and minor version number. For example, 1.2 is backward compatible with 1.1 for consumers but for providers it is incompatible.
```

Now, reading this actually something a bit odd is standing out to me. They seem to be saying that for "API consumers" minor versions must be backwards compatible, but for "API producers" minor updates are not.

Section 3.7.4 of the OSGI specification (http://docs.osgi.org/download/r8/osgi.core-8.0.0.pdf) defines "producers" and "consumers":
```
As mentioned in Semantic Versioning on page 58, there are two roles for clients of an API package:
API consumers and API providers. API consumers use the API and API providers implement the API.
```

It seems that the OSGI versioning specification is describing in better detail what the jakartaee-platform/CompatibilityRequirements are TRYING to say: API must be backwards compatible, but SPI doesn't have to be? 
If this is true, then I didnt realize there was this distinction in OSGI versioning and the current fix is legitimate as it breaks `jakarta.persistence.spi.PersistenceUnitInfo` SPI, which is permissible. My apologies.

NOTE: The reason that I am referencing OSGI versioning here when it may appear to be unrelated is because the JPA Specification produces an API bundle that has the JPA Specification version == OSGI bundle version. Therefore, to me, the two are inseparable.

**Comment by lukasj** [comment]
> it is permissible to break the SPI?

given number of vendors compared to a number of end users, there is higher chance for vendors, who are - or at least should - be members of the spec project, to come to some conclusion wrt support and eventual implementation of such change, than for users/applications as the change in SPI is on vendors' pitch and these vendors can mitigate the impact on the user apps on their end (unless user apps are doing something ... let's say not nice :-) ) - that is how I understand the motivation behind allowing this sort of changes

> Can anyone supply a document that describes what "a new version of the specification" is defined as?

I'm only aware of  https://eclipse-ee4j.github.io/jakartaee-platform/Versioning and https://wiki.eclipse.org/JakartaEE_Maven_Versioning_Rules

wrt OSGi: there is no requirement to support OSGi. To allow the usage of an API artifact there, the least effort approach is taken. While this approach does not follow semantic versioning per se, it significantly lowers the burden for the build configuration (ie no need for a bnd file or custom config of the maven-bundle-plugin) and implementations and usages, which do not want to follow that while still wanting to be usable in the OSGi environment. I've found nothing in the OSGi spec, which would mandate usage of semantic versioning. All I've found is that using semantic versioning is strongly recommended (section 3.7.3). It is also not clear how to support serviceloader in an OSGi bundle of the spec project - should this rely on HK2 resource locator, Service Loader Mediator spec or something else? If the serviceloader is not supported, does it break the OSGi spec? Note that for proper support of the semantic versioning, one needs tools. These tools should come from some Jakarta project; Jakarta spec project cannot have a dependency on non-jakarta artifact (such as OSGi specific annotations). I recommend going through the minutes of the Platform meetings to get more insight into this topic.
